src/con_utils.ts

```diff
@@ -0,0 +1,402 @@
+/*!
+ * This source file is part of the EdgeDB open source project.
+ *
+ * Copyright 2019-present MagicStack Inc. and the EdgeDB authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import * as path from "path";
+import * as url from "url";
+
+const EDGEDB_PORT = 5656;
+
+export type Address = string | [string, number];
+
+export interface NormalizedConnectConfig {
+  addrs: Address[];
+  user: string;
+  password?: string;
+  database: string;
+  connect_timeout?: number;
+  server_settings?: {[key: string]: string};
+  command_timeout?: number;
+}
+
+export interface ConnectConfig {
+  dsn?: string;
+  host?: string | string[];
+  port?: number | number[];
+  user?: string;
+  password?: string;
+  database?: string;
+  admin?: boolean;
+  timeout?: number;
+  command_timeout?: number;
+  server_settings?: any;
+}
+
+function mapParseInt(x: any): number {
+  const res = parseInt(x, 10);
+  if (isNaN(res)) {
+    throw new Error(
+      "could not convert " + JSON.stringify(x) + " to an integer"
+    );
+  }
+
+  return res;
+}
+
+export function parseConnectArguments(
+  opts: ConnectConfig = {}
+): NormalizedConnectConfig {
+  if (opts.command_timeout != null) {
+    if (typeof opts.command_timeout !== "number" || opts.command_timeout < 0) {
+      throw new Error(
+        "invalid command_timeout value: " +
+          "expected greater than 0 float (got " +
+          JSON.stringify(opts.command_timeout) +
+          ")"
+      );
+    }
+  }
+
+  return {
+    ...parseConnectDsnAndArgs(opts),
+    connect_timeout: opts.timeout,
+    command_timeout: opts.command_timeout,
+  };
+}
+
+function parseConnectDsnAndArgs({
+  dsn,
+  host,
+  port,
+  user,
+  password,
+  database,
+  admin,
+  server_settings,
+}: ConnectConfig): NormalizedConnectConfig {
+  if (dsn) {
+    const dsnHostMatch = /\/\/(.+?@)?(.*)\//.exec(dsn);
```

Comment:
I think you can use https://nodejs.org/docs/latest/api/url.html here.

Reply:
I've tried it already and it had many issues with DSN containing comma-separated list of hosts like this one `edgedb://user@host1,host2/db`. This type of "host" messed up the pathname and port results, hence why I ended up replacing it, parsing the rest of the DSN as if the host was some simple placeholder and then hand-parsing the comma-separated hosts separately.

Reply:
A comment in the code explaining this would be great.

Reply:
Yeah, good point. I've updated the comments.